src/Bicep.Core/Semantics/NameBindingVisitor.cs

```diff
@@ -1,341 +1,410 @@
+
+        private readonly IReadOnlyDictionary<SyntaxBase, LocalScope> allLocalScopes;
+
+        private readonly Stack<LocalScope> activeScopes;
+
+        public NameBindingVisitor(IReadOnlyDictionary<string, DeclaredSymbol> declarations, IDictionary<SyntaxBase, Symbol> bindings, ImmutableDictionary<string, NamespaceSymbol> namespaces, ImmutableArray<LocalScope> localScopes)
+        {
+            this.declarations = declarations;
+            this.bindings = bindings;
+            this.namespaces = namespaces;
+            this.allLocalScopes = ScopeCollectorVisitor.Build(localScopes);
+            this.activeScopes = new Stack<LocalScope>();
+        }
+
+        public override void VisitProgramSyntax(ProgramSyntax syntax)
+        {
+            base.VisitProgramSyntax(syntax);
+
+            // create bindings for all of the declarations to their corresponding symbol
+            // this is needed to make find all references work correctly
+            // (doing this here to avoid side-effects in the constructor)
+            foreach (DeclaredSymbol declaredSymbol in this.declarations.Values)
+            {
+                this.bindings.Add(declaredSymbol.DeclaringSyntax, declaredSymbol);
+            }
+
+            // include all the locals in the symbol table as well
+            // since we only allow lookups by object and not by name,
+            // a flat symbol table should be sufficient
+            foreach (var declaredSymbol in allLocalScopes.Values.SelectMany(scope => scope.Declarations))
+            {
+                this.bindings.Add(declaredSymbol.DeclaringSyntax, declaredSymbol);
+            }
+        }
+
+        public override void VisitVariableAccessSyntax(VariableAccessSyntax syntax)
+        {
+            base.VisitVariableAccessSyntax(syntax);
+
+            var symbol = this.LookupSymbolByName(syntax.Name, false);
+
+            // bind what we got - the type checker will validate if it fits
+            this.bindings.Add(syntax, symbol);
+        }
+
+        public override void VisitResourceAccessSyntax(ResourceAccessSyntax syntax)
+        {
+            base.VisitResourceAccessSyntax(syntax);
+
+            // we need to resolve which resource delaration the LHS is pointing to - and then 
+            // validate that we can resolve the name.
+            this.bindings.TryGetValue(syntax.BaseExpression, out var symbol);
+
+            if (symbol is ErrorSymbol)
+            {
+                this.bindings.Add(syntax, symbol);
+                return;
+            }
+            else if (symbol is null || symbol is not ResourceSymbol)
+            {
+                // symbol could be null in the case of an incomplete expression during parsing like `a:`
+                var error = new ErrorSymbol(DiagnosticBuilder.ForPosition(syntax.ResourceName).ResourceRequiredForResourceAccess(symbol?.Kind.ToString() ?? LanguageConstants.ErrorName));

```

Comment:
![image](https://user-images.githubusercontent.com/22460039/109705753-f6022a00-7b4c-11eb-8618-5f3fc851fa0e.png)

Should the error span be on the base expression? The root cause of the error is the fact that the base expression isn't bound to a resource symbol. On the other hand, the squiggle would cover the entire base expression, which could get kind of nasty if the expression is complicated. 

What do you think?

Reply:
what about on the colon? The error in this case is the operator you're using - which doesn't apply to the type on the LHS. Or is that too subtle?

Where would the span be for something like:
```
var test = 42
var wrong = 42.foo
```

Reply:
For the above, we do this:
![image](https://user-images.githubusercontent.com/22460039/109904422-17007300-7c52-11eb-9ef6-3fd692dae940.png)

Yeah I think `:` would be too subtle but it is also one of the possible causes of the error. I guess for consistency we could leave it where it is.

Reply:
FYI what I'm doing here is consistent with property access on a non-object type. I'm going to leave it for now unless there's a strong reason to revisit.

<img width="261" alt="image" src="https://user-images.githubusercontent.com/1430011/109904613-5cbd3b80-7c52-11eb-988f-d74c05b953b9.png">

Reply:
Sounds good to me!